### PR TITLE
[Dev] Fix DSA indexer metric accounting under Dynamic CP and CUDA Graph capture

### PR DESCRIPTION
**Pull request by buptzyb**
## Summary

Fix DSA/CSA indexer-loss metric accounting under Dynamic Context
Parallelism, pipeline parallelism, Hybrid/MTP layer layouts, and
Transformer Engine/full-iteration CUDA Graph capture.

Main counterpart: #7290

This change only affects detached metric collection and logging. It does
not change DSA/CSA attention computation or the indexer loss used for
backward.

## Problem

The previous metric tracker made several assumptions that do not hold for
current DSA/CSA configurations:

- Dynamic CP may use a different logical CP group for each packed
  microbatch, while the tracker retained only the last writer's group.
  The accumulated metric therefore depended on the packing schedule
  instead of representing the nominal global-batch mean.

- The tracker length was used as the averaging divisor even when some
  layers reused top-k results or did not own an indexer. This
  under-reported metrics for sparse indexer schedules and Hybrid/MTP
  layouts.

- Global PP/DP process groups could be incorrect for multi-module
  schedules, causing inconsistent or asymmetric collectives on ranks
  without a language model.

- CUDA Graph capture executes tracker writes. Capture-only contributions
  could leak into runtime metrics, full-iteration capture could count an
  iteration twice, and post-capture tracker resizing could leave graphs
  writing to stale storage.

## Changes

- Reduce Dynamic-CP metrics over the stable language-model DP x CP parent
  group and preserve nominal global-batch normalization.
- Apply static CP normalization on the writer stage before PP aggregation,
  followed by LM DP averaging.
- Use explicit owning language-model process groups and skip encoder-only
  ranks.
- Average over layers that actually produce indexer losses, including DSA
  top-k sharing and Hybrid/MTP layouts.
- Pre-negotiate and allocate tracker storage before TE/full-iteration CUDA
  Graph capture.
- Remove capture-only contributions while preserving storage and reduction
  metadata required by graph replay.
- Centralize metric reduction and CUDA Graph lifecycle helpers in
  `dsa_logging.py`.

The local CUDA Graph implementation is unchanged.

## Tests

- `test_dsa_integration.py`: 21 passed
- `test_cuda_graphs.py`: 75 passed, 3 skipped
- `test_full_cuda_graph.py`: 4 passed
- `test_training.py`: 7 passed
- `CHECK_ONLY=true BASE_REF=dev tools/autoformat.sh`

### CODE REVIEW

File: megatron/core/full_cuda_graph.py [hunk 1]
```diff
@@ -213,6 +213,32 @@ def __call__(self, *args, **kwargs):
...
+                prepare_dsa_tracker = pg_collection.has_language_model()
+                pp_group = (
+                    pg_collection.get_language_model_collection().pp
+                    if prepare_dsa_tracker
+                    else None
+                )
+            else:
+                pp_group = getattr(pg_collection, "pp", None)
+            if prepare_dsa_tracker:
+                _prepare_dsa_metric_tracker_for_capture(model, pp_group)
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/full_cuda_graph.py]: Fixed in 2ebdbf89b. Capture and runtime logging now both use the owning language-model process-group collection. training_log skips encoder-only MIMO ranks and passes the LM PP and DP groups explicitly, while FullCudaGraphWrapper resolves the same LM collection for capture. reduce_loss_in_tracker therefore no longer falls back to a wider global PP group on this path, keeping collective participation and tensor shapes symmetric.

File: megatron/core/transformer/cuda_graphs.py [hunk 2]
```diff
@@ -378,6 +471,11 @@ def create_cudagraphs(cls):
             )
             return
 
+        # Local capture follows one complete eager step, so the local tracker already reflects
+        # every local hybrid/MTP index. Negotiate its final size before any PP stage records a
+        # graph; otherwise the first metric reduction could grow another stage's captured tensor.
+        _prepare_dsa_metric_tracker_for_capture()
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py]: Fixed in 2ebdbf89b. All three local schedule call sites now pass model and pg_collection, so the capture scan sees every potential DSA writer. The empty-record decision is PP-global rather than rank-local: if the entire LM PP group is empty, every rank marks capture complete without DSA preparation; if any peer has a runner, every LM rank prepares the tracker and an empty local stage then marks its capture phase complete. This avoids both repeated no-op work and asymmetric PP collectives. Regression tests cover both paths.

File: megatron/core/transformer/cuda_graphs.py [hunk 3]
```diff
@@ -142,6 +142,99 @@ def _set_warmup_end():
...
+        size_tensor = torch.tensor(
+            [local_required_size], device=torch.cuda.current_device(), dtype=torch.long
+        )
+        torch.distributed.all_reduce(size_tensor, op=torch.distributed.ReduceOp.MAX, group=pp_group)
+        size = int(size_tensor.item())
+
+    if size > 0:
+        DSAIndexerLossLoggingHelper.ensure_tracker_size(size)
+    tracker["agreed_size"] = size
+    tracker["capture_prepared_size"] = size
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py]: Fixed in 2ebdbf89b. reduce_loss_in_tracker and track_indexer_metrics now accept explicit PP and DP groups, and training passes the same language-model collection used by local, TE, and full-iteration capture. Cached size provenance records both the group object and stable ordered ranks, rejects a different PP domain, and remains valid across process-group recreation with identical ranks. The direct global lookup remains only as a documented legacy fallback.

File: megatron/core/transformer/cuda_graphs.py [hunk 4]
```diff
@@ -142,6 +142,99 @@ def _set_warmup_end():
...
+    return size
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py]: Fixed in 2ebdbf89b. Zero is never published as runtime agreed_size, so a later writer renegotiates normally; capture_prepared_size may still hold zero as a capture-local sentinel. Local, TE, and full-iteration teardown now clear capture-only provenance after the relevant graphs are actually gone, while preserving runtime values, positive agreed_size, and reduction groups. Tests cover zero renegotiation and delete then recapture lifecycle.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 5]
```diff
@@ -394,18 +414,24 @@ def reduce_loss_in_tracker(num_layers: Optional[int] = None):
...
-        # Reduce indexer losses across ranks.
-        if tracker.get('reduce_group') is not None:
-            torch.distributed.all_reduce(values, group=tracker.get('reduce_group'))
-        if tracker.get('avg_group') is not None:
+        if dynamic_cp_parent_group is not None:
+            values.mul_(configured_cp_size)
             torch.distributed.all_reduce(
-                values, group=tracker['avg_group'], op=torch.distributed.ReduceOp.AVG
+                values, group=dynamic_cp_parent_group, op=torch.distributed.ReduceOp.AVG
+            )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Fixed in 2ebdbf89b. Dynamic-CP logging now validates that the parent size is divisible by configured_cp_size and, when the explicit LM DP group is available, that the parent is exactly DP size times configured CP size. A wrong parent domain now raises before reduction, with a focused regression test.

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsa_integration.py [hunk 6]
```diff
@@ -454,6 +454,169 @@ def test_indexer_loss_tracker_grows_for_mtp_layer_numbers():
...
+        )
+    )
+
+    tracker_sum = sum(
+        (local_sums * configured_cp_size).sum() / parent_size for local_sums in scheduled_local_sums
+    )
+    nominal_num_microbatches = expected_sample_sums.numel() // configured_dp_size
+
+    torch.testing.assert_close(tracker_sum / nominal_num_microbatches, expected_sample_sums.mean())
+
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply tests/unit_tests/transformer/experimental_attention_variant/test_dsa_integration.py]: Fixed in 2ebdbf89b. The arithmetic-only test was replaced with a production-path test that calls track_indexer_metrics, simulates the PP and Dynamic-CP parent collectives, and checks the configured-CP multiplier, nominal microbatch divisor, collective routing, final logged mean, and tracker reset.

File: megatron/training/training.py [hunk 7]
```diff
@@ -3478,6 +3478,11 @@ def training_log(
     # Track sparse attention indexer loss.
     if args.dsa_indexer_loss_coeff is not None and args.dsa_indexer_loss_coeff > 0:
         indexer_loss_scale = 1 / get_num_microbatches()
+        dynamic_cp_parent_group = None
+        if args.dynamic_context_parallel:
+            dynamic_cp_parent_group = getattr(pg_collection, "dp_cp", None)
+            if dynamic_cp_parent_group is None:
+                dynamic_cp_parent_group = mpu.get_data_parallel_group(with_context_parallel=True)
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/training/training.py:3710]: Fixed in 2ebdbf89b. The nominal get_num_microbatches divisor is retained intentionally, and the load-bearing sequence-packing and Dynamic-CP rationale is now documented directly above indexer_loss_scale. The training test also verifies the nominal divisor together with the LM parent-group routing.

File: megatron/core/transformer/cuda_graphs.py [hunk 8]
```diff
@@ -1763,6 +1764,11 @@ class TECudaGraphHelper:
     parameters that are covered by cudagraphs.
     """
 
+    # Subclasses that capture a disjoint module (for example, a vision encoder) must opt out:
+    # scanning their full parent model can discover language-model DSA writers whose tracker
+    # storage is not referenced by the subclass's graphs.
+    _captures_dsa_metric_tracker = True
```
Comment (kunlunl) [inline megatron/core/transformer/cuda_graphs.py]: One thing I’m a little concern about here: cuda_graphs.py is generic infrastructure, but this change makes it explicitly aware of DSA bookkeeping (dsa_logging, _captures_dsa_metric_tracker, etc.). It feels like DSA-specific logic is leaking into the CUDA graph layer, especially since VisionTECudaGraphHelper now needs to explicitly opt out.

File: megatron/core/full_cuda_graph.py [hunk 9]
```diff
@@ -216,6 +218,25 @@ def __call__(self, *args, **kwargs):
         training_str = 'training' if training else 'validation'
         curr_iteration = self.curr_iter(training_str)
         if curr_iteration == self.cuda_graph_warmup_steps:
+            pg_collection = kwargs.get('pg_collection')
```
Comment (kunlunl) [inline megatron/core/full_cuda_graph.py]: Same concern as in cuuda_graphs.py

**Review by kunlunl — APPROVED** [review]

File: megatron/training/training.py [hunk 10]
```diff
@@ -3658,21 +3684,57 @@ def training_log(
...
+        if dsa_metric_pg_collection is not None:
+            pp_group = dsa_metric_pg_collection.pp
+            dp_group = dsa_metric_pg_collection.dp
+        if args.dynamic_context_parallel:
+            if dsa_metric_pg_collection is not None:
+                dynamic_cp_parent_group = dsa_metric_pg_collection.dp_cp
+            else:
+                # Legacy training entry points without a process-group collection.
+                dynamic_cp_parent_group = mpu.get_data_parallel_group(with_context_parallel=True)
+        tracked_layers = args.num_layers + (args.mtp_num_layers or 0)
```
Comment (kunlunl) [inline megatron/training/training.py:3769]: Does this handle mtp_use_repeated_layer=True correctly? In that mode, MTP builds one shared layer and executes it mtp_num_layers times?

Comment (buptzyb) [reply megatron/training/training.py:3769]: Fixed, thanks!

File: megatron/core/transformer/cuda_graphs.py (hunk 8 above)
Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py]: Make sense. I removed almost all DSA handles except for the tracker cleaning after TE cudagraph capturing.

**Review by yaox12 — APPROVED** [review]
